### PR TITLE
installer: cluster management

### PR DESCRIPTION
**Pull request by jvatic**
screen shot 2015-04-22 at 11 39 20
screen shot 2015-04-22 at 11 40 00

### CODE REVIEW

File: installer/aws_cluster.go [hunk 1]
```diff
@@ -0,0 +1,574 @@
...
+	}()
+}
+
+func (c *AWSCluster) Delete() {
+	c.cf = cloudformation.New(c.creds, c.Region, nil)
+	go func() {
+		stackEventsSince := time.Now()
+		c.cluster.setState("deleting")
+		if err := c.fetchStack(); err != StackNotFoundError {
+			fmt.Println(err)
```
Comment (titanous) [inline installer/aws_cluster.go]: This looks like a debug print?

File: installer/aws_cluster.go [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+	c.ImageID = imageID
+	if err := c.saveField("ImageID", imageID); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *AWSCluster) fetchLatestEC2Images() ([]*release.EC2Image, error) {
+	client := &http.Client{}
+	resp, err := client.Get("https://dl.flynn.io/ec2/images.json")
```
Comment (titanous) [inline installer/aws_cluster.go]: No need, for the `client`, just do `http.Get`

File: installer/aws_cluster.go [hunk 3]
```diff
@@ -0,0 +1,574 @@
...
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.New(fmt.Sprintf("Failed to fetch list of flynn images: %s", resp.Status))
+	}
+	dec := json.NewDecoder(resp.Body)
+	manifest := release.EC2Manifest{}
+	err = dec.Decode(&manifest)
+	if err != nil {
```
Comment (titanous) [inline installer/aws_cluster.go]:
```go
if err := json.NewDecoder(resp.Body).Decode(&manifest); err != nil {
```

File: installer/aws_cluster.go [hunk 4]
```diff
@@ -0,0 +1,574 @@
...
+	return nil
+}
+
+func (c *AWSCluster) fetchLatestEC2Images() ([]*release.EC2Image, error) {
+	client := &http.Client{}
+	resp, err := client.Get("https://dl.flynn.io/ec2/images.json")
+	if err != nil {
+		return nil, err
+	}
+	if resp.StatusCode != http.StatusOK {
```
Comment (titanous) [inline installer/aws_cluster.go]:
```go
defer resp.Body.Close()
```

File: installer/aws_cluster.go (hunk 2 above)
Comment (titanous) [inline installer/aws_cluster.go]: s/resp/res

File: installer/events.go [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+		cluster: c,
+	})
+	return res.Input
+}
+
+func (c *Cluster) sendEvent(event *Event) {
+	c.installer.SendEvent(event)
+}
+
+func (c *Cluster) SendInstallLogEvent(description string) {
```
Comment (titanous) [inline installer/events.go]: This is a mouthful, perhaps shorten to `SendLog` or something like that?

File: installer/aws_cluster.go [hunk 6]
```diff
@@ -0,0 +1,574 @@
...
+func (e StackEventSort) Less(i, j int) bool {
+	return e[j].Timestamp.After(e[i].Timestamp)
+}
+
+func (c *AWSCluster) waitForStackCompletion(action string, after time.Time) error {
+	stackID := aws.String(c.StackID)
+
+	actionCompleteSuffix := "_COMPLETE"
+	actionFailureSuffix := "_FAILED"
+	actionInProgressSuffix := "_IN_PROGRESS"
```
Comment (titanous) [inline installer/aws_cluster.go]: These should all be `const` (they can stay in the method body)

File: installer/aws_cluster.go [hunk 7]
```diff
@@ -0,0 +1,574 @@
...
+	isComplete := false
+	isFailed := false
+
+	stackEvents := make([]cloudformation.StackEvent, 0)
```
Comment (titanous) [inline installer/aws_cluster.go]: Use `var` if you're not allocating space.

File: installer/cluster.go [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+		fields = append(fields, c.ID)
+		for i, ip := range c.InstanceIPs {
+			insertStmt += fmt.Sprintf("INSERT INTO instances (ClusterID, IP) VALUES ($1, $%d);\n", i+2)
+			fields = append(fields, ip)
+		}
+		q := fmt.Sprintf(`
+    %s
+    `, insertStmt)
+		_, err = tx.Exec(q, fields...)
+	case "Domain":
```
Comment (titanous) [inline installer/cluster.go]: This is weird, I think it's better served by separate methods to do `Domain` and `InstanceIPs`.

File: installer/cluster.go [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+		return err
+	}
+
+	switch field {
+	case "InstanceIPs":
+		insertStmt := ""
+		fields := make([]interface{}, 0, len(c.InstanceIPs)+1)
+		fields = append(fields, c.ID)
+		for i, ip := range c.InstanceIPs {
+			insertStmt += fmt.Sprintf("INSERT INTO instances (ClusterID, IP) VALUES ($1, $%d);\n", i+2)
```
Comment (titanous) [inline installer/cluster.go]: Why not multiple exec calls instead of appending strings?

File: installer/cluster.go [hunk 10]
```diff
@@ -0,0 +1,410 @@
...
+
+	if _, err = tx.Exec(`DELETE FROM instances WHERE ClusterID == $1`, c.ID); err != nil {
+		return
+	}
+
+	if _, err = tx.Exec(`DELETE FROM clusters WHERE ID == $1`, c.ID); err != nil {
+		return
+	}
+
+	if _, err = tx.Exec(`DELETE FROM aws_clusters WHERE ClusterID == $1`, c.ID); err != nil {
```
Comment (titanous) [inline installer/cluster.go]: Probably better to soft delete.

File: installer/http.go [hunk 11]
```diff
@@ -365,87 +157,69 @@ func (api *httpAPI) InstallHandler(w http.ResponseWriter, req *http.Request, par
...
-	api.InstallerStackMtx.Lock()
-	defer api.InstallerStackMtx.Unlock()
-	id := params.ByName("id")
-	s := api.InstallerStacks[id]
-	if s == nil {
-		w.WriteHeader(404)
+func (api *httpAPI) DeleteCluster(w http.ResponseWriter, req *http.Request, params httprouter.Params) {
+	if err := api.Installer.DeleteCluster(params.ByName("id")); err != nil {
+		if err == ClusterNotFoundError {
+			w.WriteHeader(404)
```
Comment (titanous) [inline installer/http.go]: This should be an object not found error via `Error`

File: installer/installer.go [hunk 12]
```diff
@@ -1,874 +1,263 @@
...
-		s.SendEvent(fmt.Sprintf("Using saved key pair (%s)", s.SSHKeyName))
-		return nil
-	}
-
-	s.SendEvent("Creating key pair")
-	keypair, err := sshkeygen.Generate()
+	list, err := ql.Compile(fmt.Sprintf(`
+		INSERT INTO clusters VALUES (%s);
+		INSERT INTO aws_clusters VALUES(%s);
+	`, strings.Join(clustersVStr, ", "), strings.Join(awsVStr, ", ")))
```
Comment (titanous) [inline installer/installer.go]: Just do two separate execs.

File: installer/installer.go [hunk 13]
```diff
@@ -1,874 +1,263 @@
...
-	var value string
-	for _, o := range s.Stack.Outputs {
-		if *o.OutputKey == name {
-			value = *o.OutputValue
-			break
+func (i *Installer) ClusterDeleted(id string) {
+	i.clustersMtx.Lock()
+	clusters := make([]interface{}, 0, len(i.clusters))
+	for _, c := range i.clusters {
+		idValue := reflect.Indirect(reflect.ValueOf(c)).FieldByName("ID")
```
Comment (titanous) [inline installer/installer.go]: This is sketchy, can you use an interface to do it?

File: installer/schema.go [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+	NumInstances        int64             `json:"num_instances"`
+	ControllerKey       string            `json:"controller_key,omitempty"`
+	ControllerPin       string            `json:"controller_pin,omitempty"`
+	DashboardLoginToken string            `json:"dashboard_login_token,omitempty"`
+	Domain              *Domain           `json:"domain" ql:"-"`
+	CACert              string            `json:"ca_cert"`
+	SSHKey              *sshkeygen.SSHKey `json:"-" ql:"-"`
+	SSHKeyName          string            `json:"ssh_key_name,omitempty"`
+	VpcCidr             string            `json:"vpc_cidr_block,omitempty"`
+	SubnetCidr          string            `json:"subnet_cidr_block,omitempty"`
```
Comment (titanous) [inline installer/schema.go]: s/Cidr/CIDR/

Comment (titanous) [inline installer/schema.go]: Ditto for the rest in this file.

**Comment by jvatic** [comment]
@titanous comments addressed

**Comment by jvatic** [comment]
One thing that needs addressing is the aws rate limit

File: installer/aws_cluster.go [hunk 15]
```diff
@@ -0,0 +1,605 @@
...
+	const maxBackoff = 10 * time.Second
+	backoff := 1 * time.Second
+	for {
+		err := runRequest()
+		if err != nil && err.Error() == rateExceededErrStr {
+			time.Sleep(backoff)
+			backoff = backoff * 2
+			if backoff > maxBackoff {
+				backoff = maxBackoff
+			}
```
Comment (titanous) [inline installer/aws_cluster.go]: This should eventually time out.

File: installer/events.go [hunk 16]
```diff
@@ -0,0 +1,273 @@
...
+    INSERT INTO %s VALUES(%s);
+	`, tableName, strings.Join(vStr, ", ")))
+	if err != nil {
+		return err
+	}
+	tx, err := i.db.Begin()
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(list.String(), fields...)
```
Comment (titanous) [inline installer/events.go]: Is this transaction required? There is only one statement.

Comment (titanous) [inline installer/events.go]: Looks like there are a few other similar transactions that only have one statement.

Comment (jvatic) [inline installer/events.go]: Database mutations require a transaction, but I think they are implicitly added so should be able to just call `db.Exec`.

File: installer/aws_cluster.go (hunk 15 above)
Comment (jvatic) [inline installer/aws_cluster.go]: Testing this with around 6 parallel installs, the backoff rarely exceeded `8s` (mostly `2s` and `4s`). Does a timeout after 34 seconds (5 retries) sound reasonable?

Comment (titanous) [inline installer/aws_cluster.go]: Yeah, sounds good.

File: installer/events.go (hunk 16 above)
Comment (jvatic) [inline installer/events.go]: Actually, looks like the transaction is needed.

**Comment by jvatic** [comment]
@titanous comments addressed

File: installer/aws_cluster.go [hunk 17]
```diff
@@ -0,0 +1,609 @@
...
+	c.base.installer.dbMtx.Lock()
+	defer c.base.installer.dbMtx.Unlock()
+
+	tx, err := c.base.installer.db.Begin()
+	if err != nil {
+		return err
+	}
+	_, err = tx.Exec(fmt.Sprintf(`
+  UPDATE aws_clusters SET %s = $2 WHERE ClusterID == $1
+  `, field), c.ClusterID, value)
```
Comment (titanous) [inline installer/aws_cluster.go]: Single statement transaction?

Comment (jvatic) [inline installer/aws_cluster.go]: ql gives an error for any mutation attempted outside a transaction, would it be better to make that part of the exec string?

**Comment by titanous** [comment]
LGTM (squash before merge)

**Pull request merged**